automod/engine/context.go

```diff
@@ -0,0 +1,184 @@
+package engine
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/bluesky-social/indigo/atproto/identity"
+	"github.com/bluesky-social/indigo/atproto/syntax"
+)
+
+// The primary interface exposed to rules.
+type BaseContext struct {
+	// Actual golang "context.Context", if needed for timeouts etc
+	Ctx context.Context
+	// Any errors encountered while processing methods on this struct (or sub-types) get rolled up in this nullable field
+	Err error
+	// slog logger handle, with event-specific structured fields pre-populated. Pointer, but expected to never be nil.
+	Logger *slog.Logger
+
+	engine  *Engine // NOTE: pointer, but expected never to be nil
+	effects Effects
+}
+
+type AccountContext struct {
+	BaseContext
+
+	Account AccountMeta
+}
+
+type RecordContext struct {
```

Comment:
Okay so I'm definitely not complaining if there's these two types specializing the different kinds of business context.  But, I am curious.

The effects type contains "extra" fields that only apply to record event processing and are inapplicable to account event processing... and there, we just say those stay empty/nil.  Here, we make more types.

I don't think there's anything bad in this asymmetry, but is there a reason behind it worth documenting or is it just incidental?  Do we think it'll continue to make sense?

Reply:
The symmetric more-types alternative is probably `AccountEffects` + `RecordEffects` (embeds AccountEffects), and then `AccountContext` embeds AccountEffects and `RecordContext` embeds RecordEffects.  (I can't especially think of a reason for a distinct "base" type for effects.)

The symmetric less-types alternative is just having one business `Context` type here and having RecordOp and AccountMeta fields that are documented to have a possibility of being nil depending on the, uh, context.

Reply:
My thinking on this is that the Context types are a public API, supporting mutation, and it is nice to use types make only sensical operations possible. The complexity/burden of extra types is worth it.

Effects on the other hand are mostly internal, and at least downstream of rule execution are mostly read-only. Code can just read the fields it cares about. The complexity/burden of additional types not worth it. In particular i'd like to refactor a bit more to have things like de-dupe, circuit-break, and persist mod actions be simple functions working on Effects, not split account/record versions (might want to split that big function up in other ways, like labels vs reports).

Effects is currently a public API, mostly just to allow tests to inspect it. Maybe would be less concerning if it wasn't public?